File: yc-shell-macos/Sources/YcInputServer/YcArena.swift
import Foundation

public struct ArenaCandidate: Equatable {
    public let id: UInt32
    public let text: String
}

public enum ArenaCommand: Equatable {
    case commit(text: String)
    case setComposing(text: String)
    case finishComposing
    case deleteSurrounding(before: UInt32, after: UInt32)
    case reloadKeyboard(layout: UInt32, layoutId: String)
}

public struct ArenaSnapshot: Equatable {
    public let editorId: UInt64
    public let seq: UInt64
    public let statusFlags: UInt32
    public let composing: String
    public let candidates: [ArenaCandidate]
    public let commands: [ArenaCommand]
}

public enum YcArena {
    private static let headerSize = 32
    private static let composingLen = 64
    private static let candSlotSize = 80
    private static let cmdSlotSize = 80
    private static let maxCandidates = 9
    private static let maxCommands = 4
    private static let maxTextLen = 64

    public static func parse(_ data: Data) -> ArenaSnapshot? {
        guard data.count >= headerSize else { return nil }
        let editorId = data.withUnsafeBytes { $0.load(fromByteOffset: 0, as: UInt64.self) }
        let seq = data.withUnsafeBytes { $0.load(fromByteOffset: 8, as: UInt64.self) }
        let statusFlags = data.withUnsafeBytes { $0.load(fromByteOffset: 16, as: UInt32.self) }
        let composingCount = Int(data.withUnsafeBytes { $0.load(fromByteOffset: 20, as: UInt32.self) })
        let candCount = min(Int(data.withUnsafeBytes { $0.load(fromByteOffset: 24, as: UInt32.self) }), maxCandidates)
        let cmdCount = min(Int(data.withUnsafeBytes { $0.load(fromByteOffset: 28, as: UInt32.self) }), maxCommands)

        let composing = String(data: data.subdata(in: headerSize..<(headerSize + min(composingCount, composingLen))), encoding: .utf8) ?? ""

        var candidates: [ArenaCandidate] = []
        let slotsOff = headerSize + composingLen
        for i in 0..<candCount {
            let off = slotsOff + i * candSlotSize
            guard off + candSlotSize <= data.count else { break }
            let id = data.withUnsafeBytes { $0.load(fromByteOffset: off, as: UInt32.self) }
            let textLen = min(Int(data.withUnsafeBytes { $0.load(fromByteOffset: off + 8, as: UInt32.self) }), maxTextLen)
            let text = String(data: data.subdata(in: (off + 16)..<(off + 16 + textLen)), encoding: .utf8) ?? ""
            candidates.append(ArenaCandidate(id: id, text: text))
        }

        var commands: [ArenaCommand] = []
        let cmdsOff = slotsOff + maxCandidates * candSlotSize
        for i in 0..<cmdCount {
            let off = cmdsOff + i * cmdSlotSize
            guard off + cmdSlotSize <= data.count else { break }
            let cmdType = data.withUnsafeBytes { $0.load(fromByteOffset: off, as: UInt32.self) }
            let param0 = data.withUnsafeBytes { $0.load(fromByteOffset: off + 4, as: UInt32.self) }
            let param1 = data.withUnsafeBytes { $0.load(fromByteOffset: off + 8, as: UInt32.self) }
            let textLen = min(Int(data.withUnsafeBytes { $0.load(fromByteOffset: off + 12, as: UInt32.self) }), maxTextLen)
            let text = String(data: data.subdata(in: (off + 16)..<(off + 16 + textLen)), encoding: .utf8) ?? ""
            switch cmdType {
            case 0: commands.append(.commit(text: text))
            case 1: commands.append(.setComposing(text: text))
            case 2: commands.append(.finishComposing)
            case 3: commands.append(.deleteSurrounding(before: param0, after: param1))
            case 4: commands.append(.reloadKeyboard(layout: param0, layoutId: text))
            default: break
            }
        }

        return ArenaSnapshot(editorId: editorId, seq: seq, statusFlags: statusFlags, composing: composing, candidates: candidates, commands: commands)
    }
}

File: yc-shell-android/yc-native/src/main/java/com/yc/input/native/YcArena.kt
package com.yc.input.native

import java.nio.ByteBuffer
import java.nio.ByteOrder

/** Parsed hot-path arena snapshot (mirrors yc-ffi arena_read). */
data class ArenaSnapshot(
    val editorId: Long,
    val seq: Long,
    val statusFlags: Int,
    val composing: String,
    val candidates: List<ArenaCandidate>,
    val commands: List<ArenaCommand>,
)

data class ArenaCandidate(val id: Int, val text: String)

sealed class ArenaCommand {
    data class Commit(val text: String) : ArenaCommand()
    data class SetComposing(val text: String) : ArenaCommand()
    object FinishComposing : ArenaCommand()
    data class DeleteSurrounding(val before: Int, val after: Int) : ArenaCommand()
    data class ReloadKeyboard(val layout: Int, val layoutId: String = "") : ArenaCommand()
}

object YcArena {
    private const val HEADER_SIZE = 32
    private const val COMPOSING_LEN = 64
    private const val CAND_SLOT_SIZE = 80
    private const val CMD_SLOT_SIZE = 80
    private const val MAX_CANDIDATES = 9
    private const val MAX_ARENA_COMMANDS = 4
    private const val MAX_CAND_TEXT_LEN = 64

    const val CMD_COMMIT = 0
    const val CMD_SET_COMPOSING = 1
    const val CMD_FINISH_COMPOSING = 2
    const val CMD_DELETE_SURROUNDING = 3
    const val CMD_RELOAD_KEYBOARD = 4

    fun parse(data: ByteArray): ArenaSnapshot? {
        if (data.size < HEADER_SIZE) return null
        val buf = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN)
        val editorId = buf.getLong(0)
        val seq = buf.getLong(8)
        val statusFlags = buf.getInt(16)
        val composingLen = buf.getInt(20).coerceIn(0, COMPOSING_LEN)
        val candCount = buf.getInt(24).coerceIn(0, MAX_CANDIDATES)
        val cmdCount = buf.getInt(28).coerceIn(0, MAX_ARENA_COMMANDS)

        val composing = String(data, HEADER_SIZE, composingLen, Charsets.UTF_8)

        val candidates = mutableListOf<ArenaCandidate>()
        var slotsOff = HEADER_SIZE + COMPOSING_LEN
        repeat(candCount) { i ->
            val off = slotsOff + i * CAND_SLOT_SIZE
            if (off + CAND_SLOT_SIZE > data.size) return@repeat
            val slot = ByteBuffer.wrap(data, off, CAND_SLOT_SIZE).order(ByteOrder.LITTLE_ENDIAN)
            val id = slot.getInt(0)
            val textLen = slot.getInt(8).coerceIn(0, MAX_CAND_TEXT_LEN)
            val text = String(data, off + 16, textLen, Charsets.UTF_8)
            candidates.add(ArenaCandidate(id, text))
        }

        val commands = mutableListOf<ArenaCommand>()
        val cmdsOff = slotsOff + MAX_CANDIDATES * CAND_SLOT_SIZE
        repeat(cmdCount) { i ->
            val off = cmdsOff + i * CMD_SLOT_SIZE
            if (off + CMD_SLOT_SIZE > data.size) return@repeat
            val slot = ByteBuffer.wrap(data, off, CMD_SLOT_SIZE).order(ByteOrder.LITTLE_ENDIAN)
            val cmdType = slot.getInt(0)
            val param0 = slot.getInt(4)
            val param1 = slot.getInt(8)
            val textLen = slot.getInt(12).coerceIn(0, MAX_CAND_TEXT_LEN)
            val text = String(data, off + 16, textLen, Charsets.UTF_8)
            when (cmdType) {
                CMD_COMMIT -> commands.add(ArenaCommand.Commit(text))
                CMD_SET_COMPOSING -> commands.add(ArenaCommand.SetComposing(text))
                CMD_FINISH_COMPOSING -> commands.add(ArenaCommand.FinishComposing)
                CMD_DELETE_SURROUNDING -> commands.add(ArenaCommand.DeleteSurrounding(param0, param1))
                CMD_RELOAD_KEYBOARD -> commands.add(ArenaCommand.ReloadKeyboard(param0, text))
            }
        }

        return ArenaSnapshot(editorId, seq, statusFlags, composing, candidates, commands)
    }
}
